### PR TITLE
Implement task dumps for current-thread runtime.

### PR DESCRIPTION
**Pull request by jswrenn**
## Motivation
In Java, *thread dumps* are an invaluable tool for debugging a stuck application in production. When a user requests a threaddump from a JVM, they receive text including the following information about each thread managed by the JVM:
- The thread name.
- The thread ID.
- The thread's priority.
- The thread's execution status: e.g., `RUNNABLE`, `WAITING`, or `BLOCKED`.
  This information is useful for debugging deadlocks.
- The thread's call stack.

Thread dumps provide a starting point for debugging deadlocks. A programmer can thread dump a deadlocked program, identify the `BLOCKED` threads, and inspect their call stacks to identify the resources they are blocked on.

This PR is a first step towards bringing an analogous functionality to tokio: "task dumps". A task dump is a comprehensive snapshot of the state of a Tokio application. This proposal will allow tokio services to briefly pause their work, reflect upon their internal runtime state, and comprehensively report that state to their operator.

## Solution
This PR implements a task dump mechanism for the current-thread runtime, that provides execution traces of each task. Execution traces are produced by instrumenting tokio's leaves to conditionally capture backtraces, which are then coalesced into a tree for each task.

For example, executing this program:
```rust
use std::hint::black_box;

#[inline(never)]
async fn a() {
    black_box(b()).await
}

#[inline(never)]
async fn b() {
    black_box(c()).await
}

#[inline(never)]
async fn c() {
    black_box(tokio::task::yield_now()).await
}

#[tokio::main(flavor = "current_thread")]
async fn main() {
    tokio::spawn(a());
    tokio::spawn(b());
    tokio::spawn(c());

    let handle = tokio::runtime::Handle::current();
    let dump = handle.dump();

    for (i, task) in dump.tasks().iter().enumerate() {
        let trace = task.trace();
        println!("task {i} trace:");
        println!("{trace}");
    }
}
```
Will produce an output like:
```
task 0 trace:
╼ dump::a::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:7:19
  └╼ dump::b::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:12:19
     └╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
        └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
           └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
task 1 trace:
╼ dump::b::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:12:19
  └╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
     └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
        └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
task 2 trace:
╼ dump::c::{{closure}} at /home/ubuntu/projects/tokio/examples/dump.rs:17:40
  └╼ tokio::task::yield_now::yield_now::{{closure}} at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:72:32
     └╼ <tokio::task::yield_now::yield_now::{{closure}}::YieldNow as core::future::future::Future>::poll at /home/ubuntu/projects/tokio/tokio/src/task/yield_now.rs:50:13
```

## Limitations
- Dumps of multi-thread runtimes are `unimplemented!()`
- Dumps of current-thread runtimes must be captured from within an active runtime context.
- Dumps do not yet provide important metadata, like spawn location or task name.
- Dumps do not yet include tasks in [`LocalSet`](https://docs.rs/tokio/1.27.0/tokio/task/struct.LocalSet.html)s or [`FuturesUnordered`](https://docs.rs/futures/latest/futures/stream/struct.FuturesUnordered.html).
- Execution traces cannot yet be decorated with extra information, like function arguments.

These limitations will be addressed in future PRs.

**NOTE:** The initial state of this PR *only* instruments `yield_now`. I will similarly instrument other leaf futures after this PR passes initial review.

## Related Reading
- https://github.com/tokio-rs/tokio/issues/5457
  This PR addresses this feature proposal.
- [***📄 Tack Trace Mechanisms***](https://hackmd.io/@jswrenn/SkldX98Ci)
  This document contrasts several potential approaches to tracing future state. This PR implements "📦 Backtrace the Leaves".

### CODE REVIEW

File: tokio/src/runtime/task/trace/mod.rs
Comment (jswrenn) [inline tokio/src/runtime/task/trace/mod.rs:148]: The tests are failing on MacOS because the unwinder never produces a symbol address exactly equalling `Self::leaf` — it's off by twelve bytes. I don't yet know why this occurs, or if the 12 byte offset is reasonably consistent or an artifact of codegen.

Comment (jswrenn) [reply tokio/src/runtime/task/trace/mod.rs:148]: I've opened an issue on backtrace-rs seeking further clarification: https://github.com/rust-lang/backtrace-rs/issues/520.

File: tokio/src/runtime/scheduler/current_thread.rs
Comment (Darksonn) [inline tokio/src/runtime/scheduler/current_thread.rs:407]: I would probably encapsulate this, because I don't want ref-count logic to happen outside of `src/runtime/task`.

File: tokio/src/runtime/task/trace/mod.rs
Comment (Darksonn) [inline tokio/src/runtime/task/trace/mod.rs:37]: What is the purpose of this linked tree? Is there any reason we can't just use a `Vec<*const c_void>` in the context to store this information?

File: tokio/src/runtime/scheduler/current_thread.rs
Comment (Darksonn) [inline tokio/src/runtime/scheduler/current_thread.rs:392]: Let-else syntax is not stabilized in our MSRV.

File: tokio/src/runtime/dump.rs
Comment (Darksonn) [inline tokio/src/runtime/dump.rs:10]: There appears to be a formatting issue here, but for some reason it is not caught by rustfmt?

File: tokio/src/runtime/task/trace/mod.rs
Comment (Darksonn) [inline tokio/src/runtime/task/trace/mod.rs:10]: Tokio uses the following import style:
```suggestion
use std::cell::Cell;
use std::ffi::c_void;
use std::fmt;
use std::pin::Pin;
use std::ptr::{self, NonNull};
use std::task::{self, Poll};
```

File: tokio/src/runtime/task/trace/mod.rs
Comment (jswrenn) [reply tokio/src/runtime/task/trace/mod.rs:148]: The issue occurs because:
> On macOS the function to get the address of the enclosing function of an ip address (`_Unwind_FindEnclosingFunction`) is unreliable due to compact unwind info collapsing multiple functions with identical unwind info together

...and is thus not used. Consequently, unwinding on macOS with `backtrace-rs` currently always emits the same value for symbol address and instruction pointer.

There are a few ways we might work around this (or, even, fix the issue in `backtrace-rs`), but in the short term I think we should probably just drop macOS support for taskdumping from this initial PR.

File: tokio/src/runtime/scheduler/current_thread.rs
Comment (jswrenn) [reply tokio/src/runtime/scheduler/current_thread.rs]: Resolved by https://github.com/tokio-rs/tokio/pull/5608/commits/88ef47e962cb22cc967ce502f9371fc731dc745e?

File: tokio/src/util/linked_list.rs
Comment (carllerche) [inline tokio/src/util/linked_list.rs:299]: It isn't obvious why this method uses `pop_`. It looks like iteration, why is it mutating the list to iterate??

File: tokio/src/runtime/task/trace/tree.rs
Comment (carllerche) [inline tokio/src/runtime/task/trace/tree.rs:1]: We tend only to rename imports if there is a name conflict. I was reading the code and did not recognize the `Map` and `Set` types. `HashSet` and `HashMap` let me guess the type without looking at the import list.

File: tokio/src/runtime/scheduler/current_thread.rs
Comment (carllerche) [inline tokio/src/runtime/scheduler/current_thread.rs:393]: You will most likely have to send a signal somehow.

File: tokio/src/runtime/task/state.rs
Comment (carllerche) [inline tokio/src/runtime/task/state.rs:7]: Why is this made `pub(crate)`? I don't see it used anywhere.

File: tokio/src/runtime/task/raw.rs
Comment (carllerche) [inline tokio/src/runtime/task/raw.rs:9]: Why is this made `pub(crate)`. It is used by `tokio::runtime::task::trace` which should be covered by `pub(super)`.

File: tokio/src/runtime/task/trace/mod.rs
Comment (carllerche) [inline tokio/src/runtime/task/trace/mod.rs:231]: As an initial implementation, this is fine. Before stabilizing, we should consider how this will work when many tasks spawned onto the runtime.

**Review by carllerche — COMMENTED** [review]
Thanks for doing this. I left several comments that are not required to apply before merging the PR, but we should create a "stabilization" tracking issue where we can keep track of remaining open questions.

File: tokio/src/runtime/task/state.rs
Comment (jswrenn) [reply tokio/src/runtime/task/state.rs:7]: Whoops, that fell through the cracks of an earlier revision. I'll correct it.

File: tokio/src/util/linked_list.rs
Comment (jswrenn) [reply tokio/src/util/linked_list.rs:299]: I took a stab at writing an iterator over `LinkedList` and it segfaulted — so I shifted gears to see if I could implement `for_each` completely safely instead using the existing API.

**Comment by jswrenn** [comment]
A bunch of CI jobs are stuck on "Expected — Waiting for status to be reported". Any idea why?

**Comment by Darksonn** [comment]
That's due to a CI change on master but not on your branch. If you merge master into your branch, then the issue should go away.

File: tokio/src/util/linked_list.rs
Comment (Darksonn) [reply tokio/src/util/linked_list.rs:299]: Please try this:
```rs
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        f(&mut *curr.as_ptr());
        next = T::pointers(curr).as_ref().get_next();
    }
}
```

File: .github/workflows/ci.yml
Comment (Darksonn) [inline .github/workflows/ci.yml:193]: Actually, it seems like the CI isn't running because this is invalid syntax.

File: tokio/src/util/linked_list.rs
Comment (jswrenn) [reply tokio/src/util/linked_list.rs:299]: @Darksonn `&*curr.as_ptr()` gives a `&T::Target`, but `f` consumes a `&T::Handle`.

Comment (Darksonn) [reply tokio/src/util/linked_list.rs:299]: Okay, try this:
```Rust
let mut next = self.head;

while let Some(curr) = next {
    unsafe {
        let handle = ManuallyDrop::new(T::from_raw(curr));
        f(&handle);
        next = T::pointers(curr).as_ref().get_next();
    }
}
```

**Comment by jswrenn** [comment]
I believe I've addressed all outstanding issues blocking this PR. I've also added a ["Progress" section to the feature request](https://github.com/tokio-rs/tokio/issues/5457#:~:text=rs%3A50%3A13-,Progress) to track progress on the task dump feature.

File: tokio/src/lib.rs
Comment (Darksonn) [inline tokio/src/lib.rs:496]: I would trigger this compile error even without `rt`. You aren't going to be enabling `--cfg tokio_taskdump` without knowing about it. Possibly also compile-fail on `--cfg tokio_taskdump` without rt.

File: tokio/src/runtime/task/trace/mod.rs
Comment (Darksonn) [inline tokio/src/runtime/task/trace/mod.rs:228]:
```suggestion
    local.clear();
    injection.clear();
```

File: tokio/src/runtime/handle.rs
Comment (Darksonn) [reply tokio/src/runtime/handle.rs:255]: Similarly, many of these cfgs could be reduced to a single `#[cfg(tokio_taskdump)]` since `lib.rs` contains a `compile_error!` if you have `#[cfg(tokio_taskdump)]` without the other options also set.

File: .github/workflows/ci.yml
Comment (Darksonn) [inline .github/workflows/ci.yml:511]: You don't actually set this when you run loom tests, so I wouldn't set it when compiling them either.

File: .github/workflows/ci.yml
Comment (Darksonn) [inline .github/workflows/ci.yml:1]: Is `--cfg tokio_unstable` without `--cfg tokio_taskdump` still tested?

File: tokio/src/lib.rs
Comment (jswrenn) [reply tokio/src/lib.rs]: > Possibly also compile-fail on `--cfg tokio_taskdump` without rt.

Hm, unfortunately, this causes both a [CI failure](https://github.com/tokio-rs/tokio/actions/runs/4745864579/jobs/8428808277?pr=5608#step:7:81) an a usability issue: You might just keep `--cfg tokio_unstable --cfg tokio_taskdump` in your environment (.cargo/config makes this easy to do), but *not* always want to actually use the taskdumping feature. In such scenarios, your build will fail because of the explicit `compile_error!`.

File: .github/workflows/ci.yml
Comment (jswrenn) [reply .github/workflows/ci.yml:1]: Resolved by https://github.com/tokio-rs/tokio/pull/5608/commits/b0b11602e2864b3fd7a8f73e1c0549785aceb228

File: tokio/src/runtime/handle.rs
Comment (jswrenn) [reply tokio/src/runtime/handle.rs:255]: I'm concerned that reducing these to just `#[cfg(tokio_taskdump)]` wouldn't prevent errors in these locations under odd configurations, and thus would drown out the lone explicit `compile_error` in `lib.rs`. I think Carl's suggestion is probably the way to go.

File: tokio/src/runtime/task/trace/mod.rs
Comment (nicolasmendoza) [inline tokio/src/runtime/task/trace/mod.rs:221]: Could we make the **trace_current_thread** function more readable and easier to understand? I think we should encapsulate some of the logic into separate functions to improve the function's clarity and maintainability.
Something like this (no tested)

```rust
/// Notify all tasks and return a vector of their raw pointers.
fn notify_all_tasks(owned: &OwnedTasks<Arc<current_thread::Handle>>) -> Vec<NonNull<task::RawTask>> {
    let mut tasks = vec![];
    owned.for_each(|task| {
        task.as_raw().state().transition_to_notified_for_tracing();
        tasks.push(task.as_raw());
    });
    tasks
}

/// Collect the traces for all tasks and return a vector of `Trace` objects.
fn collect_task_traces(tasks: Vec<NonNull<task::RawTask>>) -> Vec<Trace> {
    tasks
        .into_iter()
        .map(|task| {
            let ((), trace) = Trace::capture(|| task.poll());
            trace
        })
        .collect()
}

/// Trace and poll all tasks of the current_thread runtime.
pub(in crate::runtime) fn trace_current_thread(
    owned: &OwnedTasks<Arc<current_thread::Handle>>,
    local: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
    injection: &mut VecDeque<Notified<Arc<current_thread::Handle>>>,
) -> Vec<Trace> {
    local.clear();
    injection.clear();
    let tasks = notify_all_tasks(owned);
    collect_task_traces(tasks)
}
```

File: tokio/src/runtime/dump.rs
Comment (nicolasmendoza) [inline tokio/src/runtime/dump.rs:44]: One possible improvement here is to use impl Trait to simplify the definition of methods that return iterators. For example:

```suggestion
    pub fn iter(&self) -> impl Iterator<Item = &Task> + '_ {
```

**'_** means that the lifetime of the returned iterator is bound to the lifetime of the Tasks object

**Review by nicolasmendoza — COMMENTED** [review]
GTM (Looks good to me), just some comments and minor suggestions.

File: .github/workflows/ci.yml
Comment (Darksonn) [inline .github/workflows/ci.yml:407]: Since this is the test designed to test various combinations of flags, I suggest we also try with just `--cfg tokio_unstable`.
```yml
      # Try with unstable feature flags
      - name: check --feature-powerset --unstable
        run: cargo hack check --all --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
        env:
          RUSTFLAGS: --cfg tokio_unstable -Dwarnings
      # Try with unstable and taskdump feature flags
      - name: check --feature-powerset --unstable --taskdump
        run: cargo hack check --all --feature-powerset --depth 2 -Z avoid-dev-deps --keep-going
        env:
          RUSTFLAGS: --cfg tokio_unstable --cfg tokio_taskdump -Dwarnings
```

File: tokio/src/task/yield_now.rs
Comment (Darksonn) [inline tokio/src/task/yield_now.rs:56]: This PR currently only marks `yield_now` as a leaf. That's fine, but I would like more futures to be included in the first release with taskdump support.

Since we just made a Tokio release, I think we can merge this now, since we will have plenty of time to add additional functions as leaves.

**Review by Darksonn — APPROVED** [review]

**Pull request merged**
